Week 4/profixNotationSolver/src/ProfixNotationSolver.java

```diff
@@ -1,93 +1,47 @@
-    // Calculate in order from the front.
-    // After CalculateSteps() is finished, 
-    //there should be only one correct answer term in the terms stack.
-    calculateSteps();
-    
-    double answer = terms.pop();
-    if (!terms.empty()) {
-      throw new ProfixNotationInputException("Lack of operator.");
-    }
-    
-    return answer;
-  }
-
-  // Calculate the arithemetic expression in order from the front.
-  // For every atomic expression, judge its kind(digit or operation or invalid input).
-  // Depending on its type, each performs a different process.
-  // 0. digit: push the term in the stack.
-  // 1. operator: do the operation and push the answer term in the stack.
-  // 2. else: throw an exception.
-  public void calculateSteps() throws ProfixNotationInputException {
-    for (int i = 0; i < expressionLength; ++i) {
+  // Calculate the given arithemetic expression and return the answer.
+  public double calculate(String arithmeticExpression) throws Exception {
+    double answer = 0;
+    for (int i = 0; i < arithmeticExpression.length(); ++i) {
       char term = arithmeticExpression.charAt(i);
-      int termType = judgeType(term);
-
-      switch (termType) {
-        case 0:
-          pushDigit(term);
-          break;
-        case 1:
-          // There should be two operands to do the operation.
-          // If not, throw an profix notation input exception.
-          double operand2 = 0.0;
-          double operand1 = 0.0;
-          if (!terms.empty()) {
-            operand2 = terms.pop();
-            if (!terms.empty()) {
-              operand1 = terms.pop();
-            } else {
-              throw new ProfixNotationInputException("Need two operands to calculate: invalid expression.");
-            }
-          } else {
-            throw new ProfixNotationInputException("There is no operand to calculate: invalid expression.");
-          }
-          operate(term, operand1, operand2);
-          break;
-        case 2:
-          throw new ProfixNotationInputException("Terms should be either number or operators: wrong input.");
-        default:
-          assert(false);
+      if (Character.isDigit(term)) {
+        stack.push((double) Character.digit(term, 10));
+      } else if (isOperator(term)) {
+        // Push the calculated answer to the stack.
+        double operand1 = stack.pop();
+        double operand2 = stack.pop();
+        stack.push(operate(term, operand2, operand1));
+      } else {
+        throw new InvalidInputException();
       }
     }
-    return;
-  }
-
-  // Judge the character's type(digit, operator or invalid input).
-  public int judgeType (char term) {
-    if (Character.isDigit(term)) {
-      return 0;
-    } else if ((term == '+') || (term == '-') || (term == '*') || (term == '/')) {
-      return 1;
-    } else {
-      return 2;
+    answer = stack.pop();
+    if (!stack.empty()) {
+      throw new TwoOrMoreExpressionsException();
     }
```

Comment:
there are two possibilities about stack is not empty 
1) remaining operand
2) remaining operator
You might divide the case for exception.
or you can not divide the case like the current code but the exception name is too weird.